### PR TITLE
extract bounding box from NetCDF/HDF5 files and insert into the geospatial metadata block 

### PR DESCRIPTION
**Pull request by pdurbin**
insert the bounding box into the geospatial metadata block

**What this PR does / why we need it**:

Prevent users from having to manually enter bounding box values. As with FITS, we get the values from the files themselves.

**Which issue(s) this PR closes**:

- Closes #9523

**Special notes for your reviewer**:

I tried pretty hard to get the compound field code working but ended up switching to something that works, `jsonParser.parseField` which is also used by the native API.

**Suggestions on how to test this**:

Upload various NetCDF files, especially ones that have a bounding box. Here's a nice small one with a bounding box that the code tests: https://www.ncei.noaa.gov/data/international-comprehensive-ocean-atmosphere/v3/archive/nrt/ICOADS_R3.0.0_1662-10.nc

**Does this PR introduce a user interface change? If mockups are available, please link/include them here**:

Not really. The bounding box is populated. Here's what it looks like when you are just looking at (not editing) a dataset:

It seems a little weird that we don't show any labels (north, south, east, west). I assume there's an open issue for this but it's out of scope.

Screen Shot 2023-04-19 at 4 55 39 PM

The values look more reasonable when you are editing:

Screen Shot 2023-04-19 at 4 55 56 PM

**Is there a release notes update needed for this change?**:

Yes, included.

**Additional documentation**:

Included. I updated the User Guide.

### CODE REVIEW

**Comment by pdurbin** [comment]
I've been saying at standup that I need to tweak something for S3 but I was wrong. S3 works fine so in f5be4a8 I removed the FIXME and replaced it with some details.

I'll note that @JR-1991 @atrisovic and I are still discussing a bit the getStandardLongitude method in this PR. I'm happy to discus this more with a reviewer. Here's an awesome diagram Jan made and [posted](https://dataverse.zulipchat.com/#narrow/stream/376593-geospatial/topic/extract.20lat.2Flong.20.239331/near/354199150):

diagram

I'm putting this in "ready for review".

File: src/main/java/edu/harvard/iq/dataverse/ingest/metadataextraction/impl/plugins/netcdf/NetcdfFileMetadataExtractor.java [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+        float eastAsFloat;
+        float westAsFloat;
+        try {
+            westAsFloat = Float.valueOf(westAndEastLongitude.getWestLongitude());
+            eastAsFloat = Float.valueOf(westAndEastLongitude.getEastLongitude());
+        } catch (NumberFormatException ex) {
+            return null;
+        }
+        if (westAsFloat > 180 || eastAsFloat > 180) {
+            Float westStandard = westAsFloat - 360;
```
Comment (sekmiller) [inline src/main/java/edu/harvard/iq/dataverse/ingest/metadataextraction/impl/plugins/netcdf/NetcdfFileMetadataExtractor.java:154]: I think that if we're converting 0-360 to -180-180 then we need to subtract just 180 from the given value, not 360.

Comment (JR-1991) [reply src/main/java/edu/harvard/iq/dataverse/ingest/metadataextraction/impl/plugins/netcdf/NetcdfFileMetadataExtractor.java:154]: @sekmiller I was also puzzled with this at the beginning, but subtracting 360° is the correct procedure. The [prime meridian](https://oceanservice.noaa.gov/facts/longitude.html#:~:text=The%20prime%20meridian%2C%20which%20runs,the%20eastern%20and%20western%20hemispheres.) (0° longitude) needs to stay at Greenwich and thus the interval's "cycle point" where both ends meet changes by 180°. Thus, this has to be attributed for, which leads to the 360°.

That's how I understood it so far, but this illustration may explain it better:

<p align="center">
image
</p>

Comment (pdurbin) [reply src/main/java/edu/harvard/iq/dataverse/ingest/metadataextraction/impl/plugins/netcdf/NetcdfFileMetadataExtractor.java:154]: @JR-1991 thanks! Even with that explanation, it's hard to keep straight in my head.

I was relying on a tool to check that maps are the same. That is these two URLs (minus 360) show the same map:

- https://linestrings.com/bbox/#343.68,41.8,353.78,49.62
- https://linestrings.com/bbox/#-16.320007,41.8,-6.220001,49.62

Here's a screenshot of the maps (and a bit more commentary) I originally posted at https://dataverse.zulipchat.com/#narrow/stream/376593-geospatial/topic/extract.20lat.2Flong.20.239331/near/348542658

Screen Shot 2023-05-03 at 9 20 22 AM

Comment (sekmiller) [reply src/main/java/edu/harvard/iq/dataverse/ingest/metadataextraction/impl/plugins/netcdf/NetcdfFileMetadataExtractor.java:154]: @JR-1991  OK. Looking at the globe and the example location makes sense when the value is between 180 and 360, but what the code does here is if the value for either is over 180 it subtracts 360 from both so there is a potential for a situation where one val is > 180 and the other is less and in that case, say 170, and subtracting 360 results in -190 which is off the scale below -180.

File: src/main/java/edu/harvard/iq/dataverse/ingest/IngestServiceBean.java [hunk 2]
```diff
@@ -1531,72 +1654,71 @@ private void processDatasetMetadata(FileMetadataIngest fileMetadataIngest, Datas
                         // create a new compound field value and its child 
                         // 
                         DatasetFieldCompoundValue compoundDsfv = new DatasetFieldCompoundValue();
-                        int nonEmptyFields = 0;
```
Comment (pdurbin) [inline src/main/java/edu/harvard/iq/dataverse/ingest/IngestServiceBean.java:1534]: This whitespace change and below is from me removing this code and then (just) re-adding it because it seems to be working afterall! My test still passes at least. 😅

File: src/test/java/edu/harvard/iq/dataverse/ingest/metadataextraction/impl/plugins/netcdf/NetcdfFileMetadataExtractorTest.java [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+    }
+
+    @Test
+    public void testStandardLongitude() {
+        NetcdfFileMetadataExtractor extractor = new NetcdfFileMetadataExtractor();
+        // Both are over 180. Change.
+        assertEquals(new WestAndEastLongitude("-16.320007", "-6.220001"), extractor.getStandardLongitude(new WestAndEastLongitude("343.68", "353.78")));
+        // One is over 180. Change
+        assertEquals(new WestAndEastLongitude("-260.0", "-179.0"), extractor.getStandardLongitude(new WestAndEastLongitude("100", "181")));
+        // Both are under 180. No change.
```
[the remark this answers is not in the document]
Comment (pdurbin) [reply src/test/java/edu/harvard/iq/dataverse/ingest/metadataextraction/impl/plugins/netcdf/NetcdfFileMetadataExtractorTest.java:57]: Yes, I agree. I'll ping Jan for some input on this.

File: src/main/java/edu/harvard/iq/dataverse/ingest/metadataextraction/impl/plugins/netcdf/NetcdfFileMetadataExtractor.java (hunk 1 above)
Comment (JR-1991) [reply src/main/java/edu/harvard/iq/dataverse/ingest/metadataextraction/impl/plugins/netcdf/NetcdfFileMetadataExtractor.java:154]: @sekmiller thanks for the review! Indeed subtracting from both would lead to a failure. I tested it visually and in the case of $\theta_1 = [280°, 170°]$ the result should translate to $\theta_2 = [-80°, 170°]$ - Hence if a degree is under 180 it should not be subtracted. I have updated the illustration:

<p align="center">
image
</p>

I understand it in this way, that within this domain transformation only one halb of the circle is changing. Note, in 0° - 360° domain the first half is the same as in the domain of (-180°) - 180° and thus there is a 1:1 relation if degrees are within 0-180. Yet, the other half does change to another range in a more drastic way. Tipping over 180° by a degree suddenly leads you to -179° which in the other domain corresponds to 181° - Hence the 360° provides a valid transformation for this case.

It's quite complicated though 😅

File: src/test/java/edu/harvard/iq/dataverse/ingest/metadataextraction/impl/plugins/netcdf/NetcdfFileMetadataExtractorTest.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (JR-1991) [reply src/test/java/edu/harvard/iq/dataverse/ingest/metadataextraction/impl/plugins/netcdf/NetcdfFileMetadataExtractorTest.java:57]: @pdurbin just for clarification, the right hand-side extractor transforms by subtracting 360° from both?

If so, then I have to agree, this test should fail and truly evaluate to $[100°, -179°]$ due to the above [comment](https://github.com/IQSS/dataverse/pull/9541#discussion_r1185691536) and that there can't be -260° since the domain has it's lower bound at -180°. 

Given the proposed change in the calculation, the test should look like this:

```java
assertEquals(
    new WestAndEastLongitude("100.0", "-179.0"),
    extractor.getStandardLongitude(new WestAndEastLongitude("100", "181"))
);
```

What do you think?

File: src/main/java/edu/harvard/iq/dataverse/ingest/metadataextraction/impl/plugins/netcdf/NetcdfFileMetadataExtractor.java (hunk 1 above)
Comment (JR-1991) [reply src/main/java/edu/harvard/iq/dataverse/ingest/metadataextraction/impl/plugins/netcdf/NetcdfFileMetadataExtractor.java:154]: Maybe I should add to the figure, that $f(\theta)$ is defined for $\theta \in [0°, 360°]$ and if $\theta < 0°$ its already in the desired domain.

File: src/test/java/edu/harvard/iq/dataverse/ingest/metadataextraction/impl/plugins/netcdf/NetcdfFileMetadataExtractorTest.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (pdurbin) [reply src/test/java/edu/harvard/iq/dataverse/ingest/metadataextraction/impl/plugins/netcdf/NetcdfFileMetadataExtractorTest.java]: I think it's late on a Friday but I powered through and pushed some fixes around longitude at 2edb351.

Almost as importantly, I added docs! Rules in the User Guide about how users (and QA) should expect the code to behave. I'd love to hear what you (@JR-1991) and @atrisovic and @sekmiller (whom I've been bugging all day) think! Here they are:

- West Longitude and East Longitude are expected to be in the range of -180 and 180. (When using :ref:`geospatial-search`, you should use this range for longitude.)
- If West Longitude and East Longitude are both over 180 (outside the expected -180:180 range), 360 will be subtracted to shift the values from the 0:360 range to the expected -180:180 range.
- If either West Longitude or East Longitude are less than zero but the other latitude is greater than 180 (which would imply an indeterminate domain, a lack of clarity of if the domain is -180:180 or 0:360), metadata will be not be extracted.

I also added more tests. I'm happy to add more if you all think of more edge cases. I don't have real NetCDF files to test with that have wacky values. It might be a challenge for QA to find NetCDF files to exercise all the rules.

File: src/main/java/edu/harvard/iq/dataverse/ingest/metadataextraction/impl/plugins/netcdf/NetcdfFileMetadataExtractor.java (hunk 1 above)
Comment (pdurbin) [reply src/main/java/edu/harvard/iq/dataverse/ingest/metadataextraction/impl/plugins/netcdf/NetcdfFileMetadataExtractor.java:156]: I tried to address all this in 690b8a4a just now and left this comment about it: https://github.com/IQSS/dataverse/pull/9541#discussion_r1186507839

Maybe I should have resolve other conversation and gotten it down to just this one. Oh well. Please see the other comment.

**Comment by pdurbin** [comment]
I'm ready for more review so I'm removing myself from this PR. Here are the changes since the last round:

1fb2798c44 add API test for FITS file metadata extraction #9331
189c7749a7 restore original "extract from compound" code #9331
2edb351cab document and fix up the rules for longitude, add tests #9331

Yes, I was convince the code to extract from compound fields wasn't working. I put it back and now it works! Well, it was always working, I suppose. Now I'm feeding it the data it expects!

File: src/main/java/edu/harvard/iq/dataverse/ingest/metadataextraction/impl/plugins/netcdf/NetcdfFileMetadataExtractor.java (hunk 1 above)
Comment (atrisovic) [reply src/main/java/edu/harvard/iq/dataverse/ingest/metadataextraction/impl/plugins/netcdf/NetcdfFileMetadataExtractor.java:156]: I read through the thread and agree with the comments! Instead of the lines 155-173, we can check each coordinate by itself and it would work well, ie:

```
Float westStandard = westAsFloat;
Float eastStandard = eastAsFloat;

 if (westAsFloat > 180) {
             westStandard = westAsFloat - 360;
}

if (eastAsFloat > 180){
             eastStandard = eastAsFloat - 360;
}

WestAndEastLongitude updatedWeLong = new WestAndEastLongitude(westStandard.toString(), eastStandard.toString());
return updatedWeLong;
```

**Review by atrisovic — CHANGES_REQUESTED** [review]

**Comment by kcondon** [comment]
Issues found:
1. Preview for netcdf file shows permissions issue:
Screen Shot 2023-05-09 at 11 39 47 AM
Screen Shot 2023-05-09 at 11 40 02 AM

**Comment by pdurbin** [comment]
> Preview for netcdf file shows permissions issue:

Sadly, this is a known issue. It's in "sprint ready", at least:

- #9416

The fix will need to happen over in the previewers repo so there's an issue to track it there:

- https://github.com/gdcc/dataverse-previewers/issues/25

File: src/main/java/edu/harvard/iq/dataverse/ingest/IngestServiceBean.java [hunk 4]
```diff
@@ -367,7 +369,20 @@ public List<DataFile> saveAndAddFilesToDataset(DatasetVersion version,
 						} else {
 							logger.fine("Failed to extract indexable metadata from file " + fileName);
 						}
-					} else if (FileUtil.MIME_TYPE_INGESTED_FILE.equals(dataFile.getContentType())) {
+                                        } else if (fileMetadataExtractableFromNetcdf(dataFile, tempLocationPath)) {
```
Comment (ErykKul) [inline src/main/java/edu/harvard/iq/dataverse/ingest/IngestServiceBean.java:372]: This breaks direct uploads for me. When using a direct upload, the file is `unattached` and the owner is set to 'null' on line 336. However, 'fileMetadataExtractableFromNetcdf' needs the owner dataset to determine the folder inside `storageIO.open();`. This results in a `NullPointerException`, preventing adding the files to the dataset. I have only tested with the local filesystem, but from looking at the code, this probably also breaks uploading files directly to S3. Can you investigate and fix it? Thanks.

Comment (pdurbin) [reply src/main/java/edu/harvard/iq/dataverse/ingest/IngestServiceBean.java:372]: @ErykKul yikes! This has already been merged. 😬  Could you please create an issue about this?

Comment (kcondon) [reply src/main/java/edu/harvard/iq/dataverse/ingest/IngestServiceBean.java:372]: I've reproduced this in develop: direct file upload via UI fails with a grayed out screen as if waiting for a completion signal and a server log error: 
[diredct_upload_ui_err.txt](https://github.com/IQSS/dataverse/files/11517681/diredct_upload_ui_err.txt)

Note that direct file upload continues to work via API. Also, deleting the file continues to work via ui.

Comment (pdurbin) [reply src/main/java/edu/harvard/iq/dataverse/ingest/IngestServiceBean.java:372]: I can reproduce it too. I created an issue:

- #9601

Thanks, both!
